### PR TITLE
Add Hacktoberfest feature

### PR DESCRIPTION
**Pull request by burtonr**
## Description

This change adds the new feature `hacktoberfest`
The goal of this feature is to find flyby pull requests
that are considered spam.
A pull request that is a first time contributor, not signed,
and only changing markdown (md) files is assumed to be spam.
If the contributor has read the contribution guide, they
would have signed their commit, and the handler would have
no affect.
Otherwise, the pull request is closed, and an invalid
label is added indicating that this PR should not be counted

Signed-off-by: Burton Rheutan <rheutan7@gmail.com>

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/alexellis/derek/blob/master/CONTRIBUTING.md))
Discussed with Alex via Slack

## How Has This Been Tested?

Added unit tests to verify the "markdown only" function works as expected. 

Unfortunately, I was unable to deploy and test as my deployment of Derek was not able to access the secret for some reason. I've run out of time this evening to debug further.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/alexellis/derek/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

We can update the documentation once this feature passes an Alpha/Beta phase of testing

### CODE REVIEW

**Comment by burtonr** [comment]
I've built the container with this change available here: `burtonr/derek:pr-hacktoberfest`

**Comment by burtonr** [comment]
Just found out that the `fetchPullRequestCommits()` does *not* return the list of files, although it is available in the struct.
I will need to add another method for getting the files.

**Comment by burtonr** [comment]
@alexellis  I've updated the changes to include the new `fetchPullRequestFiles()` function. The updated image is here: `burtonr/derek:pr-hacktoberfest3`

Tested with a lot of support from @rgee0 :100:

Screenshot from 2019-10-03 21-05-13

To confirm this will not affect legitimate, signed, PRs:

Screenshot from 2019-10-03 21-10-26

**Comment by alexellis** [comment]
The screenshots look good. I would like to see a message from Derek explaining as briefly as possible what happened / why. This is to mitigate honest mistakes

**Comment by burtonr** [comment]
I've added a message to the handler with a link to the contributing guide as well as the Hacktoberfest "Quality Standards" here: https://hacktoberfest.digitalocean.com/details#quality-standards

> Available in the image `burtonr/derek:pr-hacktoberfest5`

Screenshot from 2019-10-04 08-28-00

**Comment by burtonr** [comment]
As a "good to know", the Hacktoberfest site does recognize the "spam" tagged by this feature

Screenshot from 2019-10-04 08-42-26

File: handler/pullRequestHandler.go [hunk 1]
```diff
@@ -211,6 +260,22 @@ func fetchPullRequestCommits(req types.PullRequestOuter, client *github.Client)
 	return commits, nil
 }
 
+func fetchPullRequestFiles(req types.PullRequestOuter, client *github.Client) ([]*github.CommitFile, error) {
+	ctx := context.Background()
+	listOpts := &github.ListOptions{
+		Page: 0,
+	}
+	commitFiles, resp, err := client.PullRequests.ListFiles(ctx, req.Repository.Owner.Login, req.Repository.Name, req.PullRequest.Number, listOpts)
```
Comment (alexellis) [inline handler/pullRequestHandler.go]: Are we sure that isn't part of the JSON already?

Comment (burtonr) [reply handler/pullRequestHandler.go]: Yes, we found that out the hard way. The check wasn't running. After looking at the request from GH, we saw that the `files` were a separate call

Comment (alexellis) [reply handler/pullRequestHandler.go]: I was wondering if we could avoid the extra API call to check files.

File: handler/pullRequestHandler.go [hunk 2]
```diff
@@ -211,6 +260,22 @@ func fetchPullRequestCommits(req types.PullRequestOuter, client *github.Client)
...
+	if err != nil {
+		log.Fatalf("Error getting files for PR %d\n%s", req.PullRequest.Number, err.Error())
+		return nil, err
+	}
+
+	fmt.Println("Rate limiting", resp.Rate)
+	resp.Body.Close()
```
Comment (alexellis) [inline handler/pullRequestHandler.go]: Please use 

```
if resp.Body!=nil {
  defer resp.Body.Close()
}
```

**Comment by alexellis** [comment]
This is looking good.

Can we show there is no regression? Does it still work with the regular pull request use-case for DCO and for labels?

Just the user guide needs updating now, in an additional commit please.

File: handler/pullRequestHandler.go (hunk 1 above)
Comment (alexellis) [reply handler/pullRequestHandler.go]: Do any files get downloaded, or is it just a list?

`fetchPullRequestFiles` makes me feel like the disk is written too. If it's not written to, can we just go with `listPullRequestFiles` or whatever you prefer?

**Comment by alexellis** [comment]
I'll merge so we can start trying it out. Please take a look at the naming question and send another PR to amend, also update the user guide.

**Comment by burtonr** [comment]
Thank you! 
I'll do those minor updates and the user guide this evening in a new PR

**Comment by alexellis** [comment]
https://github.com/alexellis/derek/releases/tag/0.9.2

**Pull request merged**